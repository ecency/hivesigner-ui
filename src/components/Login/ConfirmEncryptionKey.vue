<template>
  <form class="import-auth-key" @submit.prevent.stop="submit">
    <icon
      class="logo text-primary mx-auto mb-16"
      name="logo"
    />
    <password-form-control
      v-model="encryptionKey"
      class="mb-6"
      name="importKey"
      :label="$t('import.hs_password')"
      type="password"
      data-e2e="confirm-encryption-key-password"
    />

    <div v-for="error of errors" :key="error" class="errors text-primary pb-5" data-e2e="confirm-encryption-key-error">
      <div>*{{ error }}</div>
    </div>

    <button :disabled="disabled" class="button button-primary w-full block" data-e2e="confirm-encryption-key-login">
      {{ $t('import.login') }}
    </button>
  </form>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'nuxt-property-decorator'
import { AccountsModule, AuthModule } from '../../store'
import Icon from '../UI/Icons/Icon.vue'
import PasswordFormControl from '../UI/Form/PasswordFormControl.vue'

@Component({
  components: { PasswordFormControl, Icon }
})
export default class ConfirmEncryptionKey extends Vue {
  @Prop()
  private account!: string

  encryptionKey = ''

  private errors = []

  private get disabled (): boolean {
    return !this.encryptionKey
  }

  private async submit (): Promise<void> {
    this.errors = []
    const valid = await AccountsModule.isValidEncryptionKey({
      username: this.account,
      password: this.encryptionKey
    })

    if (!valid) {
      this.errors.push(this.$t('import.incorrect_encryption_key') as string)
      return
    }

    const keys = await AccountsModule.getEncryptedKeys({ username: this.account, encryptionKey: this.encryptionKey })
    await AuthModule.login({ username: this.account, keys })
    await AccountsModule.setSelectedAccount(this.account)

    this.$popupMessages.show('accounts.successfully_logged_in', 5000)

    this.$emit('loggedin')
  }
}
</script>

<style scoped>

</style>
