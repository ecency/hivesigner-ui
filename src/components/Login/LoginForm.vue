<template>
  <form @submit.prevent="submitForm" method="post" class="text-left">
    <label for="username">Username</label>
    <div v-if="dirty.username && !!errors.username" class="text-primary mb-2">
      {{ errors.username }}
    </div>
    <select
      id="username"
      v-model.trim="username"
      class="input-lg block mb-2"
      autocorrect="off"
      autocapitalize="none"
      autocomplete="username"
      @change="handleBlur('username')"
    >
      <option v-for="user in Object.keys(keychain)" :key="user" :value="user">
        {{ user }}
      </option>
    </select>
    <label for="password" v-if="!decrypted">
      Hivesigner password
      <span
        class="tooltipped tooltipped-n tooltipped-multiline"
        :aria-label="tooltipLoginEncryptionKey"
      >
            <span class="iconfont icon-info"/>
          </span>
    </label>
    <div v-if="dirty.key && !!errors.key" class="text-primary mb-2">
      {{ errors.key }}
    </div>
    <input
      id="password"
      v-if="!decrypted"
      v-model.trim="loginKey"
      type="password"
      autocorrect="off"
      autocapitalize="none"
      autocomplete="current-password"
      class="input-lg block mb-2"
      :class="{ 'mb-4': !error }"
      @blur="handleBlur('key')"
    />
    <div v-if="!!error" class="text-primary mb-6">{{ error }}</div>
    <button
      :disabled="submitDisabled || loading"
      type="submit"
      class="button-primary w-full block mb-2"
    >
      Login
    </button>
  </form>
</template>

<script lang="ts">
import triplesec from 'triplesec'
import { Component, Prop, Vue } from 'nuxt-property-decorator'
import { ERROR_INVALID_ENCRYPTION_KEY, TOOLTIP_LOGIN_ENCRYPTION_KEY } from '~/consts'
import { PersistentFormsModule } from '~/store'
import { signComplete } from '~/utils'
import { Authority } from '~/enums'

@Component
export default class LoginForm extends Vue {
  @Prop({
    type: Boolean,
    default: false,
  })
  private loading!: boolean

  @Prop({
    type: String,
    default: '',
  })
  private error!: string

  @Prop({
    type: String,
    default: '',
  })
  private authority!: Authority

  @Prop({
    default: () => ({}),
  })
  private keychain!: Record<string, any>

  private decrypted = true
  private dirty = {
    username: false,
    key: false,
  }

  private get errors(): Record<string, any> {
    const current: Record<string, any> = {}
    const { username, key } = this
    if (!username) {
      current.username = 'Username is required.'
    }
    if (!key && !this.dirty.key) {
      current.key = 'Hivesigner password is required.'
    }
    return current
  }

  private get username(): string {
    return PersistentFormsModule.login.username
  }

  private set username(value: string) {
    PersistentFormsModule.saveLoginUsername(value)
  }

  private get loginKey(): string {
    return PersistentFormsModule.login.key
  }

  private set loginKey(value: string) {
    return PersistentFormsModule.saveLoginKey(value)
  }

  private get tooltipLoginEncryptionKey(): string {
    return TOOLTIP_LOGIN_ENCRYPTION_KEY
  }

  private get submitDisabled(): boolean {
    return !!this.errors.username || (!!this.errors.key && !!this.dirty.key)
  }

  private get redirect(): string {
    return this.$route.query.redirect as string
  }

  public resetForm(): void {
    this.dirty = {
      username: false,
      key: false,
    }
    this.username = ''
    this.loginKey = ''
  }

  private mounted(): void {
    this.handleBlur('username')
  }

  private submitForm(): void {
    const encryptedKeys = this.keychain[this.username]
    this.$emit('loading', true)
    if (this.decrypted) {
      this.$emit('loading', false)
      const bb = Buffer.from(encryptedKeys.replace('decrypted', ''), 'hex').toString()
      this.$emit('submit', bb)
    } else {
      // @ts-ignore
      triplesec.decrypt({
          data: new triplesec.Buffer(encryptedKeys, 'hex'),
          key: new triplesec.Buffer(this.loginKey),
        }, (decryptError, buff) => {
          if (decryptError) {
            this.$emit('loading', false)
            this.$emit('error', ERROR_INVALID_ENCRYPTION_KEY)
            console.log('err', decryptError)
            return
          }
          this.$emit('submit', buff)
        })
    }
  }

  private handleBlur(name: string): void {
    this.dirty[name] = true
    if (name === 'username' && this.keychain[this.username].includes('decrypted')) {
      this.decrypted = true
      this.dirty.key = true
    } else {
      this.decrypted = false
      this.dirty.key = true
    }
  }

  private handleReject(): void {
    const requestId = this.$route.query.requestId as string
    if (requestId) {
      signComplete(requestId, 'Request canceled', null)
    }
    this.$emit('failed', false)
    this.$emit('loading', false)
    this.$emit('signature', '')
    this.$router.push('/')
  }
}
</script>

<style scoped>

</style>
