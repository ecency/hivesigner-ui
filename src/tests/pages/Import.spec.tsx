import React from "react";
import Login from "../../pages/import";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
describe("Login component Test", () => {
  Enzyme.configure({ adapter: new Adapter() });

  test("Snapshot testing", () => {
    const ImportContainer = renderer
      .create(<div className="ImportContainer" />)
      .toJSON();
    expect(ImportContainer).toMatchSnapshot();
  });

  test("each components test", () => {
    expect(shallow(<Login />).find(".ImportMenubar").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".auth_Image").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".ImportHeader").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".logo_image").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".InputContainer").length).toEqual(2);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".InputGroup").length).toEqual(2);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".CheckDiv").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".Button").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".Signuplink").length).toEqual(1);
  });
  test("each components test", () => {
    expect(shallow(<Login />).find(".importfix").length).toEqual(1);
  });
});
