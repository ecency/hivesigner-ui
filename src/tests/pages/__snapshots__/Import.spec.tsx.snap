// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component Test Snapshot testing 1`] = `
<div
  className="ImportContainer"
/>
`;
